File: mobile/ios/Sources/ServiceService.swift
import Foundation

/// Networking for the Services feature against the local Next.js API (no
/// Supabase). Mirrors `BookingService` / `HostService`: pure URLSession +
/// Codable, reading the bearer token straight from `UserDefaults` under
/// `AuthStore.tokenKey` ("qk_token") so it stays decoupled from the auth store.
///
///   GET   {base}/api/local/services                 → [Service]        (public browse)
///   GET   {base}/api/local/services/:id             → Service          (public detail)
///   POST  {base}/api/local/services                 → 201 Service      (host)
///   POST  {base}/api/local/service-requests         → 201 ServiceRequest (user subscribe)
///   GET   {base}/api/local/service-requests         → [ServiceRequest] (my subscriptions)
///   PATCH {base}/api/local/service-requests/:id     → ServiceRequest   (host confirm|reject)
///   GET   {base}/api/local/host/services            → [Service]        (host's services)
///   GET   {base}/api/local/host/service-requests    → [ServiceRequest] (host inbox)
struct ServiceService {
    static let shared = ServiceService()

    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 15
        cfg.waitsForConnectivity = true
        return URLSession(configuration: cfg)
    }()

    /// The persisted bearer token, or `nil` when browsing as a guest.
    var token: String? {
        let value = UserDefaults.standard.string(forKey: AuthStore.tokenKey)
        return (value?.isEmpty == false) ? value : nil
    }

    // MARK: - Browse (public)

    /// All published services. Public — no token required.
    func fetchServices() async throws -> [Service] {
        try await getPublic("\(Config.apiBaseURL)/api/local/services", as: [Service].self)
    }

    /// A single service's detail. Public — no token required.
    func fetchService(id: String) async throws -> Service {
        try await getPublic("\(Config.apiBaseURL)/api/local/services/\(id)", as: Service.self)
    }

    // MARK: - Create (host only)

    /// Fields the host "Add service" form collects. Sent as the POST body.
    struct NewService {
        var title: String
        var description: String
        var category: String
        var location: String
        var price: Double
        var imageURL: String
        var lat: Double?
        var lng: Double?
    }

    /// Create a service. Throws `ServiceError.forbidden` when the signed-in
    /// account isn't a host (backend 403), `ServiceError.message` for other 4xx/5xx.
    @discardableResult
    func createService(_ service: NewService) async throws -> Service {
        guard let token else { throw ServiceError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/services")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        var body: [String: Any] = [
            "title": service.title,
            "price": service.price,
        ]
        let description = service.description.trimmingCharacters(in: .whitespacesAndNewlines)
        let category = service.category.trimmingCharacters(in: .whitespacesAndNewlines)
        let location = service.location.trimmingCharacters(in: .whitespacesAndNewlines)
        let imageURL = service.imageURL.trimmingCharacters(in: .whitespacesAndNewlines)
        if !description.isEmpty { body["description"] = description }
        if !category.isEmpty { body["category"] = category }
        if !location.isEmpty { body["location"] = location }
        if !imageURL.isEmpty { body["image_url"] = imageURL }
        if let lat = service.lat, let lng = service.lng {
            body["lat"] = lat
            body["lng"] = lng
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ServiceError.message("Invalid response from the server.")
        }
        if (200...299).contains(http.statusCode) {
            return try JSONDecoder().decode(Service.self, from: data)
        }
        if http.statusCode == 401 { throw ServiceError.notSignedIn }
        if http.statusCode == 403 {
            throw ServiceError.forbidden(Self.decodeError(data) ?? "Only hosts can create services.")
        }
        throw ServiceError.message(Self.decodeError(data) ?? "Couldn't create the service (\(http.statusCode)).")
    }

    // MARK: - Subscribe (user)

    /// Subscribe to / request a service. Throws `ServiceError.notSignedIn` when
    /// there is no token, `ServiceError.message` carrying the server's `{ error }`.
    @discardableResult
    func subscribe(serviceID: String, preferredDate: String? = nil, note: String? = nil) async throws -> ServiceRequest {
        guard let token else { throw ServiceError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/service-requests")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        var body: [String: Any] = ["service_id": serviceID]
        if let preferredDate, !preferredDate.isEmpty { body["preferred_date"] = preferredDate }
        if let note = note?.trimmingCharacters(in: .whitespacesAndNewlines), !note.isEmpty {
            body["note"] = note
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ServiceError.message("Invalid response from the server.")
        }
        if (200...299).contains(http.statusCode) {
            return try JSONDecoder().decode(ServiceRequest.self, from: data)
        }
        if http.statusCode == 401 { throw ServiceError.notSignedIn }
        throw ServiceError.message(Self.decodeError(data) ?? "Subscription failed (\(http.statusCode)).")
    }

    // MARK: - My subscriptions (user)

    /// The signed-in user's service subscriptions/requests.
    func myServiceRequests() async throws -> [ServiceRequest] {
        try await getAuthed("\(Config.apiBaseURL)/api/local/service-requests", as: [ServiceRequest].self)
    }

    // MARK: - Host

    /// The host's own services.
    func hostServices() async throws -> [Service] {
        try await getAuthed("\(Config.apiBaseURL)/api/local/host/services", as: [Service].self)
    }

    /// Incoming subscription requests across all of the host's services.
    func hostServiceRequests() async throws -> [ServiceRequest] {
        try await getAuthed("\(Config.apiBaseURL)/api/local/host/service-requests", as: [ServiceRequest].self)
    }

    /// Confirm or reject a pending subscription. `action` is `confirm` or `reject`.
    @discardableResult
    func setRequestStatus(id: String, action: HostBookingAction) async throws -> ServiceRequest? {
        guard let token else { throw ServiceError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/service-requests/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "PATCH"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["status": action.rawValue])

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ServiceError.message("Invalid response from the server.")
        }
        if (200...299).contains(http.statusCode) {
            return try? JSONDecoder().decode(ServiceRequest.self, from: data)
        }
        if http.statusCode == 401 { throw ServiceError.notSignedIn }
        throw ServiceError.message(Self.decodeError(data) ?? "Couldn't update the request (\(http.statusCode)).")
    }

    // MARK: - Helpers

    /// Public GET (no auth) → decoded `T`.
    private func getPublic<T: Decodable>(_ urlString: String, as type: T.Type) async throws -> T {
        let url = URL(string: urlString)!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        // Send the token when present (harmless on public routes).
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ServiceError.message("Invalid response from the server.")
        }
        guard (200...299).contains(http.statusCode) else {
            throw ServiceError.message(Self.decodeError(data) ?? "Failed to load services (\(http.statusCode)).")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    /// Authenticated GET → decoded `T`. Maps 401 to `.notSignedIn`, 403 to `.forbidden`.
    private func getAuthed<T: Decodable>(_ urlString: String, as type: T.Type) async throws -> T {
        guard let token else { throw ServiceError.notSignedIn }

        let url = URL(string: urlString)!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ServiceError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ServiceError.notSignedIn }
        if http.statusCode == 403 {
            throw ServiceError.forbidden(Self.decodeError(data) ?? "You don't have access to that.")
        }
        guard (200...299).contains(http.statusCode) else {
            throw ServiceError.message(Self.decodeError(data) ?? "Request failed (\(http.statusCode)).")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    private static func decodeError(_ data: Data) -> String? {
        struct ErrorBody: Decodable { let error: String }
        return (try? JSONDecoder().decode(ErrorBody.self, from: data))?.error
    }
}

/// Errors surfaced to the services UI. Mirrors `BookingError` / `HostError`.
enum ServiceError: LocalizedError {
    case notSignedIn
    case forbidden(String)
    case message(String)

    var errorDescription: String? {
        switch self {
        case .notSignedIn:         return "Sign in to continue"
        case let .forbidden(text): return text
        case let .message(text):   return text
        }
    }
}

File: mobile/android/app/src/main/java/com/quickin/app/ServiceService.kt
package com.quickin.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Minimal HTTP client for the local Next.js **services** API. Mirrors [BookingService]:
 * no third-party HTTP/JSON libraries (HttpURLConnection + org.json on Dispatchers.IO),
 * and the caller supplies the bearer token (read from SharedPreferences "qk_auth" / "token").
 *
 * "Services" = standalone experiences (jet ski, diving, yacht…). A host posts a service;
 * a user "subscribes" (a service-request) → pending → the host confirms / rejects.
 *
 *   GET   {base}/api/local/services                    -> Service[]      (public browse)
 *   GET   {base}/api/local/services/:id                -> Service
 *   POST  {base}/api/local/services            (host)  -> 201 Service
 *   POST  {base}/api/local/service-requests    (user)  -> 201            (subscribe)
 *   GET   {base}/api/local/service-requests    (user)  -> ServiceRequest[]   (my subs)
 *   PATCH {base}/api/local/service-requests/:id (host) -> ServiceRequest
 *   GET   {base}/api/local/host/services       (host)  -> Service[]
 *   GET   {base}/api/local/host/service-requests (host)-> ServiceRequest[]   (host inbox)
 */
object ServiceService {

    /** Thrown so callers can distinguish "sign in to subscribe" (401) from validation (400). */
    class HttpError(val code: Int, message: String) : RuntimeException(message)

    // ---- Browse (public) ------------------------------------------------------

    /** Lists all published services for the public browse feed (no auth required). */
    suspend fun fetchServices(): List<Service> = withContext(Dispatchers.IO) {
        val text = getPublic("/api/local/services")
        parseServices(text)
    }

    /** Fetches a single service by [serviceId] (`GET /api/local/services/:id`). */
    suspend fun fetchService(serviceId: String): Service = withContext(Dispatchers.IO) {
        val text = getPublic("/api/local/services/$serviceId")
        parseService(JSONObject(text))
    }

    // ---- Subscribe (user) -----------------------------------------------------

    /**
     * Subscribes the signed-in user to [serviceId] (`POST /api/local/service-requests`).
     * Returns the created [ServiceRequest] (201). Throws [HttpError] (401 not signed in,
     * 400 on validation).
     */
    suspend fun subscribe(
        token: String,
        serviceId: String,
        note: String? = null,
        preferredDate: String? = null
    ): ServiceRequest = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("service_id", serviceId)
            if (!note.isNullOrBlank()) put("note", note.trim())
            if (!preferredDate.isNullOrBlank()) put("preferred_date", preferredDate)
        }
        val text = send("POST", token, "/api/local/service-requests", body)
        parseServiceRequest(JSONObject(text))
    }

    /** Lists the signed-in user's service subscriptions. Throws [HttpError] (401 when signed out). */
    suspend fun myServiceRequests(token: String): List<ServiceRequest> =
        withContext(Dispatchers.IO) {
            val text = get(token, "/api/local/service-requests")
            parseServiceRequests(text)
        }

    // ---- Host -----------------------------------------------------------------

    /**
     * Creates a service as the signed-in host (`POST /api/local/services`).
     * Returns the created [Service] (201). Throws [HttpError] (403 when role != host,
     * 400 on validation).
     */
    suspend fun createService(
        token: String,
        title: String,
        description: String?,
        category: String?,
        location: String?,
        price: Double,
        imageUrl: String?,
        lat: Double? = null,
        lng: Double? = null
    ): Service = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("title", title)
            if (!description.isNullOrBlank()) put("description", description.trim())
            if (!category.isNullOrBlank()) put("category", category.trim())
            if (!location.isNullOrBlank()) put("location", location.trim())
            put("price", price)
            if (!imageUrl.isNullOrBlank()) put("image_url", imageUrl.trim())
            if (lat != null && lng != null) {
                put("lat", lat)
                put("lng", lng)
            }
        }
        val text = send("POST", token, "/api/local/services", body)
        parseService(JSONObject(text))
    }

    /** The host's own services (`GET /api/local/host/services`). Throws [HttpError] (401 / 403). */
    suspend fun hostServices(token: String): List<Service> = withContext(Dispatchers.IO) {
        val text = get(token, "/api/local/host/services")
        parseServices(text)
    }

    /**
     * Service-request inbox across the host's services (`GET /api/local/host/service-requests`).
     * Throws [HttpError] (401 not signed in, 403 when the account isn't a host).
     */
    suspend fun hostServiceRequests(token: String): List<ServiceRequest> =
        withContext(Dispatchers.IO) {
            val text = get(token, "/api/local/host/service-requests")
            parseServiceRequests(text)
        }

    /**
     * Confirms or rejects a pending subscription as the host
     * (`PATCH /api/local/service-requests/:id {status:"confirm"|"reject"}`).
     * Returns the updated [ServiceRequest]. Throws [HttpError] (401 / 403 / 400).
     */
    suspend fun setRequestStatus(
        token: String,
        requestId: String,
        action: String
    ): ServiceRequest = withContext(Dispatchers.IO) {
        val body = JSONObject().apply { put("status", action) }
        val text = send("PATCH", token, "/api/local/service-requests/$requestId", body)
        parseServiceRequest(JSONObject(text))
    }

    // ---- HTTP helpers (mirror BookingService) ---------------------------------

    /** Unauthenticated GET (public browse); returns the body text or throws [HttpError]. */
    private fun getPublic(path: String): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
        }
        try {
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    /** Authenticated GET; returns the body text or throws [HttpError] on a non-2xx. */
    private fun get(token: String, path: String): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", "Bearer $token")
        }
        try {
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    /** Authenticated [method] (POST/PATCH) with a JSON body; returns the body or throws [HttpError]. */
    private fun send(method: String, token: String, path: String, body: JSONObject): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", "Bearer $token")
        }
        try {
            conn.outputStream.use { out -> out.write(body.toString().toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun readBody(conn: HttpURLConnection, code: Int): String {
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        return stream?.bufferedReader()?.use { it.readText() }.orEmpty()
    }

    private fun extractError(text: String, code: Int): String {
        val parsed = runCatching { JSONObject(text).optString("error") }.getOrNull()
        return if (!parsed.isNullOrBlank()) parsed else "Request failed ($code)"
    }

    // ---- Parsing --------------------------------------------------------------

    private fun parseServices(json: String): List<Service> {
        val arr = JSONArray(json)
        val result = ArrayList<Service>(arr.length())
        for (i in 0 until arr.length()) result.add(parseService(arr.getJSONObject(i)))
        return result
    }

    /** Parses one service object. Internal so [WishlistService] can reuse the same shape. */
    internal fun parseService(o: JSONObject): Service = Service(
        id = o.optString("id"),
        hostId = o.optString("host_id").ifBlank { null },
        hostName = o.optString("host_name").ifBlank { null },
        title = o.optString("title"),
        description = o.optString("description").ifBlank { null },
        category = o.optString("category").ifBlank { null },
        location = o.optString("location").ifBlank { null },
        price = o.optDouble("price", 0.0),
        currency = o.optString("currency").ifBlank { null },
        imageUrl = o.optString("image_url").ifBlank { null },
        lat = o.optDouble("lat").takeUnless { it.isNaN() },
        lng = o.optDouble("lng").takeUnless { it.isNaN() },
        isPublished = o.optBoolean("is_published", true)
    )

    private fun parseServiceRequests(json: String): List<ServiceRequest> {
        val arr = JSONArray(json)
        val result = ArrayList<ServiceRequest>(arr.length())
        for (i in 0 until arr.length()) result.add(parseServiceRequest(arr.getJSONObject(i)))
        return result
    }

    private fun parseServiceRequest(o: JSONObject): ServiceRequest = ServiceRequest(
        id = o.optString("id"),
        serviceId = o.optString("service_id"),
        userId = o.optString("user_id").ifBlank { null },
        status = o.optString("status").ifBlank { "pending" },
        preferredDate = o.optString("preferred_date").ifBlank { null },
        note = o.optString("note").ifBlank { null },
        requestCode = o.optString("request_code").ifBlank { null },
        serviceTitle = o.optString("service_title"),
        serviceCategory = o.optString("service_category").ifBlank { null },
        serviceImage = o.optString("service_image").ifBlank { null },
        servicePrice = o.optDouble("service_price", 0.0),
        serviceCurrency = o.optString("service_currency").ifBlank { null },
        serviceLocation = o.optString("service_location").ifBlank { null },
        hostId = o.optString("host_id").ifBlank { null },
        hostName = o.optString("host_name").ifBlank { null },
        requesterName = o.optString("requester_name").ifBlank { null },
        requesterEmail = o.optString("requester_email").ifBlank { null }
    )
}
